ebs/ebs.go

```diff
@@ -355,79 +403,192 @@ func (d *Driver) CreateVolume(req Request) error {
+				},
+			},
+		)
 		if err != nil {
 			return err
 		}
 
-		snapshotVolumeSize := *ebsSnapshot.VolumeSize * GB
-		volumeSize, err = d.getSize(opts, snapshotVolumeSize)
+		log.Printf("live availibility zones")
+
+		// Need to specify the volume and the filter for availability-zones
+		mostRecentVolume, err := d.ebsService.GetMostRecentVolume(
+			volumeName, 
+			d.DefaultDCName,
+			&ec2.Filter {
+				Name: aws.String("availability-zone"),
+				Values: liveAvailabilityZones,
+			},
+		)
 		if err != nil {
 			return err
+		} else if mostRecentVolume == nil {
+			return fmt.Errorf("GetMostRecentVolume returned nil for known volume: %s", volumeID)
 		}
-		if volumeSize < snapshotVolumeSize {
-			return fmt.Errorf("Volume size cannot be less than snapshot size %v", snapshotVolumeSize)
+
+		// Most recent volume is the current volume
+		// Check to make sure all tags are up to date
+		if *mostRecentVolume.VolumeId == volumeID {
+			volumeSize = *ebsVolume.Size * GB
+			log.Debugf("Found EBS volume %v for volume %v, update tags", volumeID, id)
+			if err := d.ebsService.AddTags(volumeID, newTags); err != nil {
+				log.Debugf("Failed to update tags for volume %v, but continue", volumeID)
+			}
+		} else {
+			mostRecentSnapshot, err := d.ebsService.GetMostRecentSnapshot(volumeName, d.DefaultDCName)
+			if err != nil {
+				return err
+			}
+
+			// Scenarios
+			// Most Recent Snapshot is nil, rebuild from VOLUME in other AZ
+			// Most Recent Volume was created after most up to date snapshot, rebuild from VOLUME in other AZ
+			// Most Recent Volume was basis for Most Recent Snapshot, rebuild from VOLUME in other AZ
+			// None of the above scenarios, Disaster Recovery and rebuild from SNAPSHOT
+			if mostRecentSnapshot == nil || (*mostRecentVolume.CreateTime).After(*mostRecentSnapshot.StartTime) || *mostRecentSnapshot.VolumeId == *mostRecentVolume.VolumeId  {
+				log.Printf("Rebuilding Volume \"%s\" from volume in AZ %s", volumeName, *mostRecentVolume.AvailabilityZone)
+				csr := &CreateSnapshotRequest {
+					VolumeID: *mostRecentVolume.VolumeId,
+					Description: fmt.Sprintf("Creating snaphot from %s", mostRecentVolume.VolumeId),
+					Tags: convertEc2TagsToMap(mostRecentVolume.Tags),
+				}
+				// Create snapshot from volume in other AZ
+				snapshotId, err := d.ebsService.CreateSnapshot(csr)
+				if err != nil {
+					return err
+				}
+
+				if err := d.ebsService.DeleteVolume(*ebsVolume.VolumeId); err != nil {
+					return err
+				}
+
+				// Build the volume in this AZ from the snapshot we just created
+				volumeID, volumeSize, err = d.buildFromSnapshot(snapshotId, opts, convertEc2TagsToMap(mostRecentVolume.Tags))
+				if err != nil {
+					return err
+				}
+			} else { 
+				// Rebuild from snapshot
+				log.Printf("Recreating volume \"%s\" from snapshot %s", volumeName, *mostRecentSnapshot.SnapshotId)
+				// Delete the volume in this AZ
+				if err := d.ebsService.DeleteVolume(*ebsVolume.VolumeId); err != nil {
```

Comment:
consider moving deleteVolume into "buildFromSnapshot" and call delete if volumeId is not nul.